### ISSUE TITLE
MacOS/Linux Support?

### ISSUE THREAD

**Issue by LewisArdern** [microsoft/restler-fuzzer#154]
Does this tool support Mac/Linux?

When I compiled the tool and ran the quick-start.py I saw in the error:
* python3 restler-quick-start.py --api_spec_path $REDACTED/data/file.json --restler_drop_dir $REDACTED/restler-fuzzer-7.2.0/restler_bin
*  System.Exception: Platform not supported

The documentation only references Windows, so it would be nice to know if windows is the only intended target.

**Comment by PatGod** [comment]
Please see the main README.md https://github.com/microsoft/restler-fuzzer :

RESTler was designed to run on 64-bit machines with Windows or Linux.

**Comment by LewisArdern** [comment]
Sure, but all the documentation is for Windows.

**Comment by marina-p** [comment]
Hello,

It should work on Linux, but we have only been testing on Ubuntu (restler-quick-start.py is running in the build pipeline).  Could you please share more about the system and setup you have?  

Correct, MacOS isn't supported.

Thanks,

Marina

**Comment by LewisArdern** [comment]
I was trying on MacOS to start with, I'll switch over to Ubuntu and go from there. 

If I did not want to use reslter-quick-start and manually invoke the process separately e.g. compile is there examples of how I may do that for linux?

**Comment by marina-p** [comment]
TutorialDemoServer.md shows the separate phases, though it's currently Windows-specific.  I'll go through it and send a PR in the next day or so to add Linux commands.

I believe the only things that need to be changed in that tutorial for Linux are the path to the executable (```./restler/Restler```) and the paths in the arguments.

**Comment by alexbrahastoll** [comment]
Hey @marina-p , I have just tried to run RESTler on its fuzz-lean mode inside an Ubuntu Docker Container and I have run into some strange issues. RESTler fails to find the grammar or the dictionary file (I suppose), even though the files clearly exist.

I am commenting here in this thread because I used the tutorial you are mentioning as a guide (adapting it to Linux as I went).

Below, the error happening when I try to run RESTler, followed by `cat` commands showing the paths are right and the files do exist:

```
root@1286a3eb89f1:~/restler-fuzzer-bin/restler# ./Restler fuzz-lean --grammar_file /root/restler-fuzzer/restler_working_dir/Compile/grammar.py --dictionary_file /root/restler-fuzzer/restler_working_dir/Compile/dict.json --settings /root/restler-fuzzer/restler_working_dir/Compile/engine_settings.json --no_ssl
Log share was not specified.  Logs will not be uploaded.
Starting task FuzzLean...

ERROR: Task FuzzLean failed.  Exception: System.ComponentModel.Win32Exception (2): No such file or directory
   at System.Diagnostics.Process.ForkAndExecProcess(String filename, String[] argv, String[] envp, String cwd, Boolean redirectStdin, Boolean redirectStdout, Boolean redirectStderr, Boolean setCredentials, UInt32 userId, UInt32 groupId, UInt32[] groups, Int32& stdinFd, Int32& stdoutFd, Int32& stderrFd, Boolean usesTerminal, Boolean throwOnNoExec)
   at System.Diagnostics.Process.StartCore(ProcessStartInfo startInfo)
   at System.Diagnostics.Process.Start()
   at Microsoft.FSharpLu.Diagnostics.Process.startProcessAsync@241-28.Invoke(Unit unitVar)
   at Microsoft.FSharp.Control.AsyncPrimitives.CallThenInvoke[T,TResult](AsyncActivation`1 ctxt, TResult result1, FSharpFunc`2 part2) in F:\workspace\_work\1\s\src\fsharp\FSharp.Core\async.fs:line 398
   at Microsoft.FSharp.Control.AsyncPrimitives.unitAsync@589.Invoke(AsyncActivation`1 ctxt) in F:\workspace\_work\1\s\src\fsharp\FSharp.Core\async.fs:line 589
   at Microsoft.FSharp.Control.AsyncPrimitives.unitAsync@589.Invoke(AsyncActivation`1 ctxt) in F:\workspace\_work\1\s\src\fsharp\FSharp.Core\async.fs:line 589
   at Microsoft.FSharp.Control.AsyncPrimitives.unitAsync@589.Invoke(AsyncActivation`1 ctxt) in F:\workspace\_work\1\s\src\fsharp\FSharp.Core\async.fs:line 589
   at Microsoft.FSharp.Control.AsyncPrimitives.unitAsync@589.Invoke(AsyncActivation`1 ctxt) in F:\workspace\_work\1\s\src\fsharp\FSharp.Core\async.fs:line 589
   at Microsoft.FSharp.Control.Trampoline.Execute(FSharpFunc`2 firstAction) in F:\workspace\_work\1\s\src\fsharp\FSharp.Core\async.fs:line 109

Task FuzzLean failed.
Collecting logs...
root@1286a3eb89f1:~/restler-fuzzer-bin/restler# cat /root/restler-fuzzer/restler_working_dir/Compile/grammar.py
""" THIS IS AN AUTOMATICALLY GENERATED FILE!"""
from __future__ import print_function
import json
from engine import primitives
from engine.core import requests
from engine.errors import ResponseParsingException
from engine import dependencies
req_collection = requests.RequestCollection([])
# Endpoint: /incidents, method: Post
request = requests.Request([
    primitives.restler_static_string("POST "),
    primitives.restler_static_string("/"),
    primitives.restler_static_string("incidents"),
    primitives.restler_static_string(" HTTP/1.1\r\n"),
    primitives.restler_static_string("Accept: application/json\r\n"),
    primitives.restler_static_string("Host: http://localhost:4567\r\n"),
    primitives.restler_refreshable_authentication_token("authentication_token_tag"),
    primitives.restler_static_string("\r\n"),

],
requestId="/incidents"
)
req_collection.add_request(request)

# Endpoint: /incidents/{incident_id}, method: Get
request = requests.Request([
    primitives.restler_static_string("GET "),
    primitives.restler_static_string("/"),
    primitives.restler_static_string("incidents"),
    primitives.restler_static_string("/"),
    primitives.restler_fuzzable_int("1"),
    primitives.restler_static_string(" HTTP/1.1\r\n"),
    primitives.restler_static_string("Accept: application/json\r\n"),
    primitives.restler_static_string("Host: http://localhost:4567\r\n"),
    primitives.restler_refreshable_authentication_token("authentication_token_tag"),
    primitives.restler_static_string("\r\n"),

],
requestId="/incidents/{incident_id}"
)
req_collection.add_request(request)
root@1286a3eb89f1:~/restler-fuzzer-bin/restler# cat /root/restler-fuzzer/restler_working_dir/Compile/dict.json
{
  "restler_fuzzable_string": [
    "fuzzstring"
  ],
  "restler_fuzzable_string_unquoted": [],
  "restler_fuzzable_datetime": [
    "6/25/2019 12:00:00 AM"
  ],
  "restler_fuzzable_datetime_unquoted": [],
  "restler_fuzzable_uuid4": [
    "903bcc44-30cf-4ea7-968a-d9d0da7c072f"
  ],
  "restler_fuzzable_uuid4_unquoted": [],
  "restler_fuzzable_int": [
    "0",
    "1"
  ],
  "restler_fuzzable_number": [
    "0.1",
    "1.2"
  ],
  "restler_fuzzable_bool": [
    "true"
  ],
  "restler_fuzzable_object": [
    "{}"
  ],
  "restler_custom_payload": {},
  "restler_custom_payload_unquoted": {},
  "restler_custom_payload_uuid4_suffix": {}
}root@1286a3eb89f1:~/restler-fuzzer-bin/restler# cat /root/restler-fuzzer/restler_working_dir/Compile/engine_settings.json
{
  "per_resource_settings": {}
}
root@1286a3eb89f1:~/restler-fuzzer-bin/restler#
```

I am able to run restler-quick-start.py successfully.

If you find it helpful, I can create another issue reporting my whole experience so far trying to run RESTler inside an Ubuntu Docker container (I am a macOS user, that is why I am resorting to using a Docker container).

Thanks!

**Comment by marina-p** [comment]
Hi @alexbrahastoll,

This is most likely a failure to find python on the path.  RESTler should issue a different error message for each of the arguments to the file (e.g. if you try with an invalid path to the grammar, you should see an error message rather than a crash).

Can you please check if 'python' is on the path and the right version before you invoke ./Restler?

We have an upcoming feature to specify the full path to python on the command line, which will also help troubleshoot such issues, and we'll improve the error handling as part of implementing that issue.

**Comment by stishkin** [comment]
@alexbrahastoll  - we are working closely with RESTler team to make it easier to integrate RESTler into CI/CD pipelines.
The idea of the project to run RESTler in Azure. But I just merged a PR that will run same configurations locally. 
https://github.com/microsoft/rest-api-fuzz-testing/tree/main/cli
raft_local.py - is the file that enables that

Also there is official Microsoft RESTler Docker container that we maintain (using Alpine Linux):
https://hub.docker.com/_/microsoft-restlerfuzzer-restler

**Comment by wuhkuh** [comment]
> Also there is official Microsoft RESTler Docker container that we maintain (using Alpine Linux):
> https://hub.docker.com/_/microsoft-restlerfuzzer-restler

I switched from a custom dockerfile to this one (mcr.microsoft.com/restlerfuzzer/restler:v7.2.0), which also consistently raises [this error](https://github.com/microsoft/restler-fuzzer/issues/154#issuecomment-794638118) when I invoke `restler test <options>`, or `dotnet <path-to-Restler.dll> test <options>`. Interestingly enough, `restler compile` works just fine in both images (custom and official).

**Comment by marina-p** [comment]
Hello @wuhkuh, thank you for reporting this.  We will investigate.

(The compile step does not invoke python, only the test/fuzz steps, so it is likely a problem with finding python on the path.)

**Comment by marina-p** [comment]
PR #158 should resolve the issue in the official RESTler container and provide a workaround for different Python installations.

**Comment by marina-p** [comment]
@alexbrahastoll @wuhkuh Please let us know if the latest RESTler commit resolves this for you.

**Comment by wuhkuh** [comment]
Yes! That seemed to have fixed the error. It was indeed caused by an improper installation of Python 3. Thank you!

**Comment by alexbrahastoll** [comment]
@stishkin Thanks, I will take a look at the image you have mentioned.

@marina-p I have Python 3.8.2 installed and correctly configured in the PATH. I will check out the updated version of RESTler in a couple days and I let you know if it solved the issue I reported above. Thank you for the fast reply.

**Comment by taivnn9** [comment]
Hi All,

**Comment by alexbrahastoll** [comment]
@marina-p I have upgraded RESTler and I have also added pip and the libraries `requests` and `applicationinsights` as part of the steps of the Dockerfile I am working with. I added pip and the libraries after taking a look at the Dockerfile for the container mentioned by @stishkin. Are these mandatory requirements for RESTler? If so, it would be nice to add them to the list of requirements in the project's documentation.

I was able to run RESTler now, but I am not sure if what solved the issue I was having was the new version, the Python libraries I mentioned or both.

A bit out of topic, but let me make a quick question: which OpenAPI versions are supported by RESTler? I suppose this is mentioned somewhere, but I searched a bit and I was not able to find this information. I had some issues with an OpenAPI 3 spec and was wondering if RESTler supports or not this latest version of the OpenAPI specification.

Thanks!
